### PR TITLE
Cython 3 and numpy 2

### PR DESCRIPTION
**Pull request by connorjward**
# Description

Fixes https://github.com/firedrakeproject/firedrake/issues/3081

It will also allow us to update our PETSc fork as petsc4py now requires cython 3.

### CODE REVIEW

File: scripts/firedrake-install [hunk 1]
```diff
@@ -983,14 +982,10 @@ def run_pip_install(pipargs):
     # subprocesses wrote out.
     # Particularly important for debugging petsc fails.
     with environment(**blas):
-        pipargs = ["-vvv"] + pipargs
+        pipargs = ["-v"] + pipargs
```
Comment (JDBetteridge) [inline scripts/firedrake-install:985]: :joy:

**Comment by connorjward** [comment]
@Ig-dolci FYI I am having another stab at this now. If we use the release candidate for numpy 2.0 then we can hopefully land Cython 3 and numpy at the same time.

**Comment by connorjward** [comment]
This is quite the pain, now [loopy is also having issues](https://github.com/inducer/loopy/issues/835).

File: scripts/firedrake-install [hunk 2]
```diff
@@ -1347,8 +1334,9 @@ def build_and_install_torch():
         raise InstallError("CUDA installation is not available on MacOS.")
     if sys.version_info >= (3, 10):
         run_pip_install(["typing_extensions"])
-    extra_index_url = ["--extra-index-url", "https://download.pytorch.org/whl/cpu"] if args.torch == "cpu" else []
-    run_pip_install(["torch"] + extra_index_url)
+    # FIXME: use nightly to see if that makes numpy 2.0 happy
```
Comment (connorjward) [inline scripts/firedrake-install:1337]: Should really be
```
pip install torch==2.3.0 --index-url https://download.pytorch.org/whl/test/cpu
```
as detailed in https://dev-discuss.pytorch.org/t/pytorch-release-2-3-0-final-rc-is-available/1995

2.3.0 is supposed to be released 24/04.

**Comment by francesco-ballarin** [comment]
Just to double check: are the numpy changes here backward compatible with 1.x, or not?

cython 3.x is somewhat widespread now, while it will still take a while to get numpy 2.x out (at least, on installations that do not use firedrake-install)

**Comment by connorjward** [comment]
> Just to double check: are the numpy changes here backward compatible with 1.x, or not?
> 
> cython 3.x is somewhat widespread now, while it will still take a while to get numpy 2.x widely used (at least, on installations that do not use firedrake-install)

I think so. The errors fixed here were mostly due to numpy dropping support for deprecated API.

**Comment by francesco-ballarin** [comment]
Will this be closed in favor of #3546?

**Comment by connorjward** [comment]
> Will this be closed in favor of #3546?

Yes.

**Pull request closed without merging**